### PR TITLE
hyper-v: KVP test suite port from older lisa

### PR DESCRIPTION
**Pull request by mbivolan**

### CODE REVIEW

File: Testscripts/Linux/KVP-AddKeyValues.sh [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ICA_TESTRUNNING="TestRunning"      # The test is running
+ICA_TESTCOMPLETED="TestCompleted"  # The test completed successfully
+ICA_TESTABORTED="TestAborted"      # Error during setup of test
+ICA_TESTFAILED="TestFailed"        # Error while performing the test
+
+CONSTANTS_FILE="constants.sh"
+
+
+
+dos2unix utils.sh
```
Comment (iamshital) [inline Testscripts/Linux/KVP-AddKeyValues.sh]: Remove `dos2unix utils.sh`. File will be converted to unix before upload.

File: Testscripts/Linux/KVP-BasicChecks.sh [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+    if [ $? -ne 0 ]; then
+        LogMsg "ERROR: Failed to install lsof"
+        UpdateSummary "ERROR: Failed to install lsof"
+        SetTestStateAborted
+        exit 0
+    fi
+}
+
+dos2unix utils.sh
```
Comment (iamshital) [inline Testscripts/Linux/KVP-BasicChecks.sh]: Remove dos2unix utils.sh. File will be converted to unix before upload.

File: Libraries/Test-Framework.psm1 [hunk 3]
```diff
@@ -454,10 +446,15 @@ function Run-Test {
         }
     } else {
         if ($testPlatform.ToUpper() -eq "HYPERV") {
-            Apply-HyperVCheckpoint -VMnames $AllVMData.RoleName -CheckpointName "ICAbase"
-            $AllVMData = Check-IP -VMData $AllVMData
-            Set-Variable -Name AllVMData -Value $AllVMData -Scope Global
-            LogMsg "Public IP found for all VMs in deployment after checkpoint restore"
+            if ($CurrentTestData.applyCheckpoint -eq "False") {
```
Comment (iamshital) [inline Libraries/Test-Framework.psm1]: Please put `applyCheckpoint` into `AdditionalHWConfig.HyperVApplyCheckpoint` to make it consistent.
Example: https://github.com/LIS/LISAv2/blob/master/XML/TestCases/PerformanceTests.xml#L6

File: XML/TestCases/KVP.xml [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        <testScript>SQM-Basic.ps1</testScript>
+        <files></files>
+        <setupType>SingleVM</setupType>
+        <testParameters>
+            <param>TC_COVERED=SQM-01</param>
+        </testParameters>
+        <Platform>HyperV</Platform>
+        <Category>KVP</Category>
+        <Area>KVP</Area>
+        <Tags>sqm</Tags>
```
Comment (juhlee-microsoft) [inline XML/TestCases/KVP.xml]: is it worth to have separated tag? All others are kvp.

File: Testscripts/Windows/TCUtils.ps1 [hunk 5]
```diff
@@ -0,0 +1,2441 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (lizzha) [inline Testscripts/Windows/TCUtils.ps1]: Can you follow the PowerShell naming convention that we agreed?
1.	Method name, use the fashion of verb-noun based names, UpperCamelCase, such as Test-Object. We can use Get-Verb to see what verbs are available.
2.	Global Variables: UpperCamelCase, such as UserName
3.	Local variable: lowerCamelCase, such as localFileName
4.	Const: all upper cases, such as: SCRIPT_VERSION
5.	Parameter name: UpperCamelCase

**Comment by mbivolan** [comment]
All the review comments were addressed and appended in the original commmits:
- @iamshital:
    - removed dos2unix from sh scripts
    - changed applyCheckpoint param into HyperVApplyCheckpoint
- @lizzha:
    - replaced TCUtils.ps1 with testcase-utils.psm1 (contains only the methods required by the tests in this PR)
    - refactored all the scripts following the conventions you specified
- @juhlee-microsoft:
    - changed sqm tag name to kvp

The tests in this PR were run against the following images:
- Hyper-V:
    - CentOS 7.5: Everything PASS
    - Debian 8.10: Only KVP-TestRescing FAIL
    - Ubuntu 18.04: Only KVP-TestRescing FAIL
- Azure:
    - RHEL 7.5: KVP-BasicChecks FAIL
    - Ubuntu 16: KVP-BasicChecks PASS

File: XML/TestCases/KVP.xml (hunk 4 above)
Comment (juhlee-microsoft) [reply XML/TestCases/KVP.xml]: acked

**Comment by iamshital** [comment]
@mbivolan Can you please send the test logs in email?

File: Libraries/CommonFunctions.psm1 [hunk 6]
```diff
@@ -2573,4 +2573,228 @@ function CreateTestResultObject()
...
+        [String] $SSHPort,
+        [String] $Username,
+        [String] $Password
+    )
+    
+    $check1 = $true
+    $check2 = $true
+    
+    .\Tools\plink.exe -C -pw $Password -P $SSHPort $Username@$Ipv4 "ls -l /sbin/init | grep systemd"
+    if ($? -ne "True") {
```
Comment (ader1990) [inline Libraries/CommonFunctions.psm1]: use lastexitcode var

Comment (ader1990) [reply Libraries/CommonFunctions.psm1]: this applies to all the future occurrences

File: Libraries/CommonFunctions.psm1 [hunk 7]
```diff
@@ -2573,4 +2573,228 @@ function CreateTestResultObject()
...
+        $check2 = $false
+    }
+
+    return ($check1 -and $check2)
+}
+
+function Get-VMFeatureSupportStatus {
+    <#
+    .Synopsis
+        Check if VM supports one feature or not.
```
Comment (ader1990) [inline Libraries/CommonFunctions.psm1]: a feature

File: Libraries/CommonFunctions.psm1 [hunk 8]
```diff
@@ -2573,4 +2573,228 @@ function CreateTestResultObject()
...
+    .Description
+        Check if VM supports one feature or not based on comparison of curent kernel version with feature
+        supported kernel version. If the current version is lower than feature supported version, return false, otherwise return true.
```
Comment (ader1990) [inline Libraries/CommonFunctions.psm1]: line too long

File: Libraries/CommonFunctions.psm1 [hunk 9]
```diff
@@ -2573,4 +2573,228 @@ function CreateTestResultObject()
...
+    
+    param (
+        [String] $Ipv4,
+        [String] $SSHPort,
+        [String] $Username,
+        [String] $Password,
+        [String] $SupportKernel
+    )
+    
+    echo y | .\Tools\plink.exe -C -pw $Password -P $SSHPort $Username@$Ipv4 'exit 0'
```
Comment (ader1990) [inline Libraries/CommonFunctions.psm1]: @iamshital @chvalean is there a reason not to use ssh.exe from portable git? or a manual build? 
we would not be required to always execute this line to accept the key

File: Testscripts/Linux/KVP-AddKeyValues.sh [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    echo "Error: unable to source utils.sh!"
+    exit 0
+}
+
+UtilsInit
+
+#
+# Delete any summary.log files from a previous run
+#
+rm -f ~/summary.log
```
Comment (ader1990) [inline Testscripts/Linux/KVP-AddKeyValues.sh]: reuse utils.sh and not use summary.log directly

**Pull request merged**